File: java/LeadDatabase/Leads/GetLeadsByFilterType.java
package dev.marketo.samples.Leads;

import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.NoSuchElementException;
import java.util.Scanner;

import javax.net.ssl.HttpsURLConnection;

import com.eclipsesource.json.JsonObject;

//the Java sample code on dev.marketo.com uses the minimal-json package
//minimal-json provides easy and fast representations of JSON
//for more information check out https://github.com/ralfstx/minimal-json

public class MultipleLeads {
	public String marketoInstance = ;//Replace this with the host from Admin Web Services
	public String marketoIdURL = marketoInstance + "/identity";	
	public String clientId = ;	//Obtain from your Custom Service in Admin>Launchpoint
	public String clientSecret = ;	//Obtain from your Custom Service in Admin>Launchpoint
	public String idEndpoint = marketoIdURL + "/oauth/token?grant_type=client_credentials&client_id=" + clientId + "&client_secret=" + clientSecret;
	
	public Integer batchSize = null;
	public String nextPageToken = null;
	public String[] fields = null;
	public String filterType; //required, must match the REST API name of a Marketo field
	public String[] filterValues;
	
	public static void main(String[] args) {
		MultipleLeads leads = new MultipleLeads();
		leads.filterType = "email"; //add a filterType, email
		leads.filterValues = new String[]{"kelkington@marketo.com"}; //add an email address to search for
		String result = leads.getData();
		System.out.println(result);
		}
    //Make request
    private String getData() {
        String data = null;
        try {
        	//Assemble the URL to retrieve data from
        	StringBuilder endpoint = new StringBuilder(marketoInstance + "/rest/v1/leads.json?access_token=" + getToken() + "&filterType=" + filterType + "&filterValues=" + csvString(filterValues));
        	//append optional params
            if (batchSize != null){
            	endpoint.append("&batchSize=" + batchSize);
            }
            if (nextPageToken != null){
            	endpoint.append("&nextPageToken=" + nextPageToken);
            }
            if (fields != null){
            	endpoint.append("&fields=" + csvString(fields));
            }
            URL url = new URL(endpoint.toString());
            HttpsURLConnection urlConn = (HttpsURLConnection) url.openConnection();
            urlConn.setRequestMethod("GET");
            urlConn.setRequestProperty("accept", "text/json");
            int responseCode = urlConn.getResponseCode();
            if (responseCode == 200) {
                InputStream inStream = urlConn.getInputStream();
                data = convertStreamToString(inStream);
            } else {
                data = "Status:" + responseCode;
            }
        } catch (MalformedURLException e) {
            System.out.println("URL not valid.");
        } catch (IOException e) {
            System.out.println("IOException: " + e.getMessage());
            e.printStackTrace();
        }

        return data;
    }
	//takes an array of fields as strings and concatenates them with alternating commas to use in a URL param
	private String csvString(String[] fields) {
		StringBuilder fieldCsv = new StringBuilder();
    	for (int i = 0; i < fields.length; i++){
    		fieldCsv.append(fields[i]);
    		if (i + 1 != fields.length){
    			fieldCsv.append(",");
    		}
    	}
		return fieldCsv.toString();
	}
	public String getToken(){
		String token = null;
		try {
			URL url = new URL(idEndpoint);
			HttpsURLConnection urlConn = (HttpsURLConnection) url.openConnection();
			urlConn.setRequestMethod("GET");
            urlConn.setRequestProperty("accept", "application/json");
            int responseCode = urlConn.getResponseCode();
            if (responseCode == 200) {
                InputStream inStream = urlConn.getInputStream();
                Reader reader = new InputStreamReader(inStream);
                JsonObject jsonObject = JsonObject.readFrom(reader);
                token = jsonObject.get("access_token").asString();
            }else {
                throw new IOException("Status: " + responseCode);
            }
		} catch (MalformedURLException e) {
			e.printStackTrace();
		}catch (IOException e) {
            e.printStackTrace();
        }
		return token;
	}
    private String convertStreamToString(InputStream inputStream) {

        try {
            return new Scanner(inputStream).useDelimiter("\\A").next();
        } catch (NoSuchElementException e) {
            return "";
        }
    }
}

File: c#/LeadDatabase/Leads/GetLeadsByFilterType.cs
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net;
using System.Text;
using System.Threading.Tasks;

namespace Samples
{
    class MultipleLeads
    {
        private String host = "CHANGE ME"; //host of your marketo instance, https://AAA-BBB-CCC.mktorest.com
        private String clientId = "CHANGE ME"; //clientId from admin > Launchpoint
        private String clientSecret = "CHANGE ME"; //clientSecret from admin > Launchpoint
        public String filterType; //field to filter by, required
        public String[] filterValues; //values to filter on, required
        public int batchSize;//max 300, default 300
        public String[] fields;//array of field names to retrieve
        public String nextPageToken;//paging token
        /*
        public static void Main(String[] args)
        {
            MultipleLeads leads = new MultipleLeads();
            leads.filterType = "email";
            leads.filterValues = new String[] { "kelkington@marketo.com" };
            String result = leads.getData();
            Console.WriteLine(result);
            while (true)
            {

            }
        }
        */
        public String getData()
        {
            StringBuilder url = new StringBuilder(host + "/rest/v1/leads.json?access_token=" + getToken() + "&filterType=" + filterType + "&filterValues=" + csvString(filterValues));
            if (fields != null)
            {
                url.Append("&fields=" + csvString(fields));
            }
            if (batchSize > 0 && batchSize < 300)
            {
                url.Append("&batchSize=" + batchSize);
            }
            if (nextPageToken != null)
            {
                url.Append("&nextPageToken=" + nextPageToken);
            }
            HttpWebRequest request = (HttpWebRequest)WebRequest.Create(url.ToString());
            request.ContentType = "application/json";
            request.Accept = "application/json";
            HttpWebResponse response = (HttpWebResponse)request.GetResponse();
            Stream resStream = response.GetResponseStream();
            StreamReader reader = new StreamReader(resStream);
            return reader.ReadToEnd();
        }

        private String getToken()
        {
            String url = host + "/identity/oauth/token?grant_type=client_credentials&client_id=" + clientId + "&client_secret=" + clientSecret;
            HttpWebRequest request = (HttpWebRequest)WebRequest.Create(url);
            request.ContentType = "application/json";
            HttpWebResponse response = (HttpWebResponse)request.GetResponse();
            Stream resStream = response.GetResponseStream();
            StreamReader reader = new StreamReader(resStream);
            String json = reader.ReadToEnd();
            //Dictionary<String, Object> dict = JavaScriptSerializer.DeserializeObject(reader.ReadToEnd);
            Dictionary<String, String> dict = JsonConvert.DeserializeObject<Dictionary<String, String>>(json);
            return dict["access_token"];
        }
        private String csvString(String[] args)
        {
            StringBuilder sb = new StringBuilder();
            int i = 1;
            foreach(String s in args){
                if (i < args.Length){
                    sb.Append(s + ",");
                }else{
                    sb.Append(s);
                }
                i++;
            }
            return sb.ToString();

        }
    }
}using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net;
using System.Text;
using System.Threading.Tasks;

namespace Samples
{
    class MultipleLeads
    {
        private String host = "CHANGE ME"; //host of your marketo instance, https://AAA-BBB-CCC.mktorest.com
        private String clientId = "CHANGE ME"; //clientId from admin > Launchpoint
        private String clientSecret = "CHANGE ME"; //clientSecret from admin > Launchpoint
        public String filterType; //field to filter by, required
        public String[] filterValues; //values to filter on, required
        public int batchSize;//max 300, default 300
        public String[] fields;//array of field names to retrieve
        public String nextPageToken;//paging token
        /*
        public static void Main(String[] args)
        {
            MultipleLeads leads = new MultipleLeads();
            leads.filterType = "email";
            leads.filterValues = new String[] { "kelkington@marketo.com" };
            String result = leads.getData();
            Console.WriteLine(result);
            while (true)
            {

            }
        }
        */
        public String getData()
        {
            StringBuilder url = new StringBuilder(host + "/rest/v1/leads.json?access_token=" + getToken() + "&filterType=" + filterType + "&filterValues=" + csvString(filterValues));
            if (fields != null)
            {
                url.Append("&fields=" + csvString(fields));
            }
            if (batchSize > 0 && batchSize < 300)
            {
                url.Append("&batchSize=" + batchSize);
            }
            if (nextPageToken != null)
            {
                url.Append("&nextPageToken=" + nextPageToken);
            }
            HttpWebRequest request = (HttpWebRequest)WebRequest.Create(url.ToString());
            request.ContentType = "application/json";
            request.Accept = "application/json";
            HttpWebResponse response = (HttpWebResponse)request.GetResponse();
            Stream resStream = response.GetResponseStream();
            StreamReader reader = new StreamReader(resStream);
            return reader.ReadToEnd();
        }

        private String getToken()
        {
            String url = host + "/identity/oauth/token?grant_type=client_credentials&client_id=" + clientId + "&client_secret=" + clientSecret;
            HttpWebRequest request = (HttpWebRequest)WebRequest.Create(url);
            request.ContentType = "application/json";
            HttpWebResponse response = (HttpWebResponse)request.GetResponse();
            Stream resStream = response.GetResponseStream();
            StreamReader reader = new StreamReader(resStream);
            String json = reader.ReadToEnd();
            //Dictionary<String, Object> dict = JavaScriptSerializer.DeserializeObject(reader.ReadToEnd);
            Dictionary<String, String> dict = JsonConvert.DeserializeObject<Dictionary<String, String>>(json);
            return dict["access_token"];
        }
        private String csvString(String[] args)
        {
            StringBuilder sb = new StringBuilder();
            int i = 1;
            foreach(String s in args){
                if (i < args.Length){
                    sb.Append(s + ",");
                }else{
                    sb.Append(s);
                }
                i++;
            }
            return sb.ToString();

        }
    }
}